### ISSUE TITLE
DUE PWM to VFD Support

### ISSUE THREAD

**Issue by DeanCavaliere** [synthetos/g2#452]
I'm not sure if this has been covered anywhere; but I have set up a VFD configured to be externally controlled. I'm using UGS to control the spindle PWM pin on the DUE (Pin 11). I have the vfd fully configured for my particular spindle and I can't seem to get it to work with the DUE. 
The DUE's PWM signal goes to a 0-10v analog board and then to the vfd. 
Is there extra steps to setting up a spindle like this that I am missing? Is there a way to set the max RPM for the spindle? (In GRBL this is setting $30)
I noticed that I can control the pwm frequency and duty cycle using {"p1":{ "freq":#, "pof": #}}
But when using UGS's spindle input box under the state tab, I do not see any change in the PWM signal on pin 11.

**Comment by justinclift** [comment]
Interesting.  This is something I want to do as well, for more powerful spindle motors than the "stock" ones I've been using so far.

Hopefully one of the other developers or a Community member has already gone through this process and gotten it working. :smile:

**Comment by karoria** [comment]
Have you tested analog voltage output after your pwm to analog converter. It should change according to live rpm set in ugs. Also, make sure that your "run" setting is also configured to command externally and wired accordingly and you give signal for that through a relay (it should be isolated when in idle mode and conducted when in run mode). 
Most of the people have issues with getting in run mode. Once run mode is activated then it moves the spindle according to analog voltage input.

**Comment by DeanCavaliere** [comment]
I just re-wired everything and took some time to make sure what I was doing was correct. I relaunched UGS and tested. 
![image](https://user-images.githubusercontent.com/49283626/74103166-e49c4900-4b17-11ea-8e54-75fe441806fd.png)
Here is the outputs in console when I run spindle reset override, spindle++ override, and a s### via the state tab.
I tested the PWM pin and analog pin and saw no changes in the magnitude, frequency, or duty cycle as I did all of these commands. 
Is there additional configuration that needs to be done in UGS? I could not find any information on it.

**Comment by DeanCavaliere** [comment]
Update: Using the {"p1"...} command, I can change pin 11's pwm and I can now see the frequency and rpm change. There was some extra set up required on the VFD which I got sorted last night and will document/make a video about for those seeking to do a similar set up. 
The m3 s### command seems to work; BUT, I'm not seeing the desired control that I would like. m5 results in a duty cycle of 10% which yields around 2500 RPM, m3 s1000 to 20000 seems to yield 3000-5000 RPM. I am fairly certain that my VFD is set properly as per guides and manufacturers documentation for external control and the wiring is correct.

After a decent nights rest, I've suddenly realized what 'p1' is.
{"p1":{"frq":100,"csl":7900,"csh":12800,"cpl":0.13,"cph":0.17,"wsl":7900,"wsh":12800,"wpl":0.13,"wph":0.17,"pof":0.1}}

P1 is used to set the PWM pin 11's defaults ('duh'). With pof being the default duty cycle. I'm not sure what csl/h, cpl/h, wsl/h, and wpl/h is and would love some clarification on what these mean. I have reason to believe that they are set points for RPM and duty cycle but am very unsure and do not want to start changing things without knowing what they are. 
Please let me know if you have any insight! Thank you so much!

**Comment by karoria** [comment]
Ok. It seems g2core configuration issue. I have not touched g2core for almost a year now, but i can remember one of the branches at that time was giving default lowest spindle pwm duty cycle as 0.13 and highest was 0.17 making it difficult to notice any change in spindle rpm. Then i found that setting somewhere and changed it to 0 and 1 respectively and then it worked like a charm.

**Comment by justinclift** [comment]
Sounds like progress is being made, and the full solution is *nearly* there...

**Comment by karoria** [comment]
OMG, it is still present.. check above cpl, cph, wpl and wph. Change it to 0 (for low) and 1 (for high) respectively and set your spindle rpm as per spindle's specs. And I am sure, you will get it working.

**Comment by karoria** [comment]
> OMG, it is still present.. check above cpl, cph, wpl and wph. Change it to 0 (for low) and 1 (for high) respectively and set your spindle rpm as per spindle's specs. And I am sure, you will get it working.

Sorry, I didn't read your previous post fully. You were already near to this...

**Comment by DeanCavaliere** [comment]
got the spindle up and running after changing those cpl/h... variables. However, on power off and restart, these values are wiped. Is there a way to flash the desired values to the board to not have to deal with rescaling every time??
Or is there a way to initially run a set of commands on start up/ connection with the DUE?

And the JSON command to set up the desired PWM is as follows (assuming 24000 RPM is your spindles max)
{"p1":{"frq":100,"csl":0,"csh":24000,"cpl":0.0,"cph":1.0,"wsl":0,"wsh":24000,"wpl":0.0,"wph":1.0,"pof":0.0}}

**Comment by justinclift** [comment]
> Or is there a way to initially run a set of commands on start up/ connection with the DUE?

Not with the Due board.  Instead, you'll need to compile the g2core code with the updated setting values (generally changed in [settings_default.h](https://github.com/synthetos/g2/blob/c684edcd38a18e4014860a9bf1ace1dc3b37285f/g2core/settings/settings_default.h)), then flash the Due with that.

There are step by step instructions for compiling and flashing, if that's useful?

&nbsp; &nbsp; https://github.com/synthetos/g2/wiki/Getting-Started-with-g2core-Development#compiling-the-code

**Comment by DeanCavaliere** [comment]
Update: 
I got the Atmel project and re-configured the settings_default.h as @justinclift mentioned and the UGS output is shown below.
![image](https://user-images.githubusercontent.com/49283626/74245850-aaff4580-4cb1-11ea-8ab8-c4251e7fcbee.png)

However, I have one more question. As shown in the image below, "fbc" is set to settings_shapeoko2.h. 
![image](https://user-images.githubusercontent.com/49283626/74246113-fd406680-4cb1-11ea-97ce-4ef1af43b923.png)
What is this doing? where do I configure this in atmel? It doesn't seem to be affecting anything but I'd like confirmation of what the heck it is if possible.
Thank you all so much for your help! @justinclift @karoria

**Comment by justinclift** [comment]
Interesting.  `fbc` shows the firmware configuration used (eg the name of the settings file).

Personally I'm used to compiling under Linux, by running this as the command line:

```
$ make PLATFORM=DUE BOARD=gShield
```

That doesn't use the Atmel IDE though.  On (say) Windows + the Atmel IDE, I'm not personally sure how to do it.  I think I tried it out once a few years ago, but have no real recollection of the steps. :man_shrugging:

If your newly compiled firmware is picking up the changes you made though (eg to `settings_default.h`), then I'd call this "good enough" and not be overly worried. :wink:

If the changes aren't being seen though, then updating `settings_shapeoko2.h` instead of `settings_default.h` would probably be the right approach. :smile:

**Comment by giseburt** [comment]
I wouldn't recommend changing settings_default.h - those are literally the defaults, and anything changed there and set in the used settings file will be overridden, which could get very confusing. In this case, as indicated by fbc, the file settings_shapeoko2.h is being used. That is decided by the boards.mk file (for the CONFIG chosen on the make command line) or overridden on the make command line as SETTINGS_FILE.

In this case the easiest thing is to change settings_shapeoko2.h. If there's a setting in settings_default.h that you need to change, copy it to settings_shapeoko2.h and change it there. As for order of settings in the file, it mostly doesn't matter, as long as they don't refer to each other. If they do, what's being referred to has to be defined before it's being used. Otherwise, order isn't important, just whatever makes the most sense for readability - and if you're not publishing it, that's whatever the heck  you want :-).

-Rob

**Comment by justinclift** [comment]
@DynamicDumbA55 How'd you go with this? :smile:

**Comment by DeanCavaliere** [comment]
@justinclift So, I was able to get the PWM set by changing the values in settings_default.h.  However, atmel studios somehow decided to use settings_shapeoko2.h. So, I think I have some more poking around to do to see how to change this.... 
And by looking around these preset configs, I've figured out how to set up a fourth motor as a mirror of the Y axis (wooohoo). 

So in short, atmel confuses the heck out of me; but it seems to be working :)

And sorry for the delay to reply, been a but hectic on my end!

**Comment by justinclift** [comment]
No worries about the delay, I'm several tasks behind at the moment myself. :wink:

And yeah, the main thing is you got it working, so all good. :grin:

**Comment by Mgilbride** [comment]
So, a question here. Did you just update the spindle speed max to 24000? I am not seeing anything else to update there.

**Comment by DeanCavaliere** [comment]
Oh it's been an eternity since I looked at this project. I briefly scrolled through the settings file and for our application, we changed quite a few variables. Spindle max was one of them, but we also had to configure jerk, velocity, feedrate [maybe don't alter these until you decide your machine is moving too fast/slow for your liking], motors 1-4 as per the specs of the motors, and for PWM input we changed the P1_PWM settings [~ line 1050 - 1079] to match the specs of the VFD. 

Hopefully that answers your question?
